include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,515 @@
+#ifndef TRITON_TOOLS_LINEARLAYOUT_H
+#define TRITON_TOOLS_LINEARLAYOUT_H
+
+#include <cstdint>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/STLExtras.h"
+
+namespace mlir::triton {
+
+// # High-level overview of linear layouts
+//
+// The idea for linear layouts is due to Adam P. Goucher.
+//
+// In Triton, a linear layout (LL) is a function that maps from a "hardware
+// location" to a "logical tensor index".
+//
+// For example, suppose we have a 2D tensor T stored in GPU registers.  T's
+// layout is the function that, given a "hardware location" tuple of (thread-id,
+// warp-id), returns an index (x,y) into T.  In other words, if L(t,w) = (x,y)
+// is our linear layout func, then thread t in warp w contains the value T[x,y].
```

Comment:
```suggestion
// is our linear layout func, then the value T[x,y] is stored in a register from thread t in warp w.
```

The notion that a thread "contains" a value seems not well defined to me, so just an alternate possible phrasing here.

Reply:
Thanks, will include in next upload